core/common/src/format/DateTimeFormat.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019-2023 JetBrains s.r.o. and contributors.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime.format
+
+import kotlinx.datetime.*
+import kotlinx.datetime.internal.format.*
+import kotlinx.datetime.internal.format.parser.*
+
+/**
+ * A format for parsing and formatting date-time-related values.
+ */
+public sealed interface DateTimeFormat<T> {
+    /**
+     * Formats the given [value] into a string, using this format.
+     */
+    public fun format(value: T): String
+
+    /**
+     * Formats the given [value] into the given [appendable] using this format.
+     */
+    public fun <A : Appendable> formatTo(appendable: A, value: T): A
+
+    /**
+     * Parses the given [input] string as [T] using this format.
+     *
+     * @throws IllegalArgumentException if the input string is not in the expected format or the value is invalid.
+     */
+    public fun parse(input: CharSequence): T
+
+    /**
+     * Parses the given [input] string as [T] using this format.
+     *
+     * @return the parsed value, or `null` if the input string is not in the expected format or the value is invalid.
+     */
+    public fun parseOrNull(input: CharSequence): T?
+
+    public companion object {
+        /**
+         * Produces Kotlin code that, when pasted into a Kotlin source file, creates a [DateTimeFormat] instance that
+         * behaves identically to [format].
+         *
+         * The typical use case for this is to create a [DateTimeFormat] instance using a non-idiomatic approach and
+         * then convert it to a builder DSL.
+         */
+        public fun formatAsKotlinBuilderDsl(format: DateTimeFormat<*>): String = when (format) {
+            is AbstractDateTimeFormat<*, *> -> format.actualFormat.builderString(allFormatConstants)
+        }
+    }
+}
+
+/**
+ * The style of padding to use when formatting a value.
+ */
+public enum class Padding {
+    /**
+     * No padding.
+     */
+    NONE,
+
+    /**
+     * Pad with zeros.
+     */
+    ZERO,
+
+    /**
+     * Pad with spaces.
+     */
+    SPACE
+}
+
+internal fun Padding.toKotlinCode(): String = when (this) {
```

Comment:
May be worth to move to `toKotlinCode.kt`

Reply:
I tried to stick to the principle that the code expected to change should be kept together. For example, `DayOfWeekNames.toKotlinCode` is directly below `DayOfWeekNames`, the field specifications are right next to the directives involving the field, etc. WDYT?

Reply:
ok